src/Microsoft.AspNetCore.SignalR.Core/Internal/HubMethodDescriptor.cs

```diff
@@ -0,0 +1,149 @@
+            MethodExecutor = methodExecutor;
+            ParameterTypes = methodExecutor.MethodParameters.Select(p => p.ParameterType).ToArray();
+            Policies = policies.ToArray();
+
+            NonAsyncReturnType = (MethodExecutor.IsMethodAsync)
+                ? MethodExecutor.AsyncResultType
+                : MethodExecutor.MethodReturnType;
+
+            if (IsObservableType(NonAsyncReturnType, out var observableItemType))
+            {
+                IsObservable = true;
+                StreamReturnType = observableItemType;
+            }
+            else if (IsChannelType(NonAsyncReturnType, out var channelItemType))
+            {
+                IsChannel = true;
+                StreamReturnType = channelItemType;
+            }
+        }
+
+        private Func<object, CancellationToken, IAsyncEnumerator<object>> _convertToEnumerator;
+
+        public ObjectMethodExecutor MethodExecutor { get; }
+
+        public IReadOnlyList<Type> ParameterTypes { get; }
+
+        public Type NonAsyncReturnType { get; }
+
+        public bool IsObservable { get; }
+
+        public bool IsChannel { get; }
+
+        public bool IsStreamable => IsObservable || IsChannel;
+
+        public Type StreamReturnType { get; }
+
+        public IList<IAuthorizeData> Policies { get; }
+
+        private static bool IsChannelType(Type type, out Type payloadType)
+        {
+            var channelType = type.AllBaseTypes().FirstOrDefault(t => t.IsGenericType && t.GetGenericTypeDefinition() == typeof(ChannelReader<>));
+            if (channelType == null)
+            {
+                payloadType = null;
+                return false;
+            }
+
+            payloadType = channelType.GetGenericArguments()[0];
+            return true;
+        }
+
+        private static bool IsObservableType(Type type, out Type payloadType)
+        {
+            var observableInterface = IsIObservable(type) ? type : type.GetInterfaces().FirstOrDefault(IsIObservable);
+            if (observableInterface == null)
+            {
+                payloadType = null;
+                return false;
+            }
+
+            payloadType = observableInterface.GetGenericArguments()[0];
+            return true;
+
+            bool IsIObservable(Type iface)
+            {
+                return iface.IsGenericType && iface.GetGenericTypeDefinition() == typeof(IObservable<>);
+            }
+        }
+
+        public IAsyncEnumerator<object> FromObservable(object observable, CancellationToken cancellationToken)
+        {
+            if (_convertToEnumerator == null)
+            {
+                _convertToEnumerator = CompileConvertToEnumerator(FromObservableMethod);
+            }
+
+            return _convertToEnumerator.Invoke(observable, cancellationToken);
+        }
+
+        public IAsyncEnumerator<object> FromChannel(object channel, CancellationToken cancellationToken)
+        {
+            if (_convertToEnumerator == null)
```

Comment:
Everything else on this type is immutable, do we want to compile this delegate up front or lazily? Are there any thread safety concerns or will it just do more work.

Reply:
No thread safety issues. Just more work if there are multiple requests for a streamable method on startup.

I think it should be lazy. Decreases startup time and overall memory usage by only compiling methods that are used.

One optimization could be to place them in a concurrent dictionary and share compiled methods between method descriptors. Each method that returns `IObservable<int>` will compile its own copy right now.

Reply:
`ObjectMethodExecutor` isn't lazy and it also doesn't share compiled methods.

I'd be curious to see how MVC uses `ObjectMethodExecutor` - whether it creates instances for it up front for every action method or whether creates them on first invoke. Right now SignalR creates them up front.

Reply:
`ControllerActionInvokerCache.GetCachedResult` creates `ObjectMethodExecutor` the first time an action is called. IMO SignalR should do the same. Compiling a lot of methods unnecessarily will hurt startup time and memory usage.

Reply:
Agreed, we should do the same as MVC here.